tsl/src/nodes/gapfill/gapfill_exec.c

```diff
@@ -640,15 +640,15 @@ gapfill_advance_timestamp(GapFillState *state)
 	{
 		case DATEOID:
 			next = DirectFunctionCall2(date_pl_interval,
-									   DateADTGetDatum(state->next_timestamp),
-									   IntervalPGetDatum(state->gapfill_interval));
+									   DateADTGetDatum(state->gapfill_start),
+									   IntervalPGetDatum(state->next_offset));
```

Comment:
Previous implementation depended on previous timestamp to calculate the next one. This becomes a problem in case of variable bucket sizes like months when the bucket lands on the last day of the month (which is a different day every time).

The computation you suggest is the same one I'm doing here but instead of maintaining the position (n), I maintain the offset by incrementing it for every generated bucket.

My implementation seemed simpler in terms of calculating the next timestamp so that's why I went with it.